### ISSUE TITLE
What is Types of Ports?

### ISSUE THREAD

**Issue by walteweiss** [corpnewt/USBMap#37]
The documentation, list 9 says it this way:

> Go into the P. Edit & Create `USBMap.kext` menu and change the types to match the physical port types and toggle which ports (up to 15) you want to preserve.

Okay, so in the terminal, while using the tool, it says:

- Change types using this formula `T:1,2,3,4,5:t` where t is the type

But… **what is the type?** 

I assume the type is written with big and small letters, so it's why it's `Tt` there. Or not?

Is it `S` or `A`?

- `S:1,2,3,4,5:s` (`S` for speed), 
- or `A:1,2,3,4,5:a` (`A` for USB-A)

Is it `2` and `3`?

- or should it be `2:1,2,3,4,5:2` (`2` for USB-2)
- and then `3:1,2,3,4,5:3` (`3` for USB-3)

Is it `HS` and `SS`?

- or is it `HS:1,2,3,4,5:hs` (`HS` for High Speed)
- and then `SS:1,2,3,4,5:ss` (`SS` for Super Speed)

Is it `H` plus `s` and `S` plus `s`?

- or is it `H:1,2,3,4,5:s` (`H` and `s` for High Speed)
- and then `S:1,2,3,4,5:s` (`S` and `s` for Super Speed)

In my listing I have it written this way:

```
[#]  1. HS01 | AppleUSB20XHCIPort |  1 (01000000) | 14100000 | Type 3
    Front-Left-USB2
...
[#] 18. SS04 | AppleUSB30XHCIPort | 19 (13000000) | 14200000 | Type 3
    Back-Down-6-USB3
```

So what type is here? `HS` and `SS` in the beginning of the line, or the number after each `Type` word? I have `3` for all the ports I have.

I assume this step is needed, that's why it's written. But I have absolutely no idea what does it do here, why won't I skip it and how to make it. 

Would be great to explain.

**Comment by walteweiss** [comment]
I see there is an issue #1 here, which reflects a similar subject, but there is no answer to what should be instead of those `T` and `t`.

**Comment by walteweiss** [comment]
I found that when I use `USBMap.command` and `USBMapInjectorEdit.command` the interfaces are slightly different. And in the latter, there is a kind of help when you press `T`, which shows it this way:

```
0: Type A connector
1: Mini-AB connector
2: ExpressCard
3: USB 3 Standard-A connector
4: USB 3 Standard-B connector
5: USB 3 Micro-B connector
6: USB 3 Micro-AB connector
7: USB 3 Power-B connector
8: Type C connector - USB2-only
9: Type C connector - USB2 and SS with Switch
10: Type C connector - USB2 and SS without Switch
11 - 254: Reserved
255: Proprietary connector

Per the ACPI 6.2 Spec.
```

So 
- for most popular Type A it's 0 
- and `3` for USB3-A
- And `8`,`9`,`10` for Type C connector (-only, with Switch, without Switch respectively).

Ok. That suggests that there's a number. So, is my latest assumption correct, and I need to use letter `T`, then ports, then `0` or `3`?

E.g.

-  `T:1,2,3,4,5:0` and `T:6,7,8,9,15:3`

**Comment by walteweiss** [comment]
Yes, it works when I did `T:1,3,4,5,6,10,11:0`. Also it works when later I did `T:7:0`, so it works like a toggle.

The documentation needs to be clarified, as it’s unclear on most steps, and only trials and errors help. I may try to contribute when I understand this better myself.

**Comment by walteweiss** [comment]
So, summarizing the long post. I think it may be closed, as my issue was solved, by trials and errors. 

But before doing that I would recommend to add some info for the type somewhere (at least Readme.md) 

Maybe add some more info in the tool, like that:

>Change types using this formula `T:1,2,3,4,5:t` where `t` is the type (0 to 10, or 255, see documentation)

But where is that documentation?

or 

>Change types using this formula `T:1,2,3,4,5:t` where `t` is the type (0 to 10, or 255), e.g. `T:1:0` for USB-A port or `T:16:3` for USB-3 port.

----

Also, one question is still relevant:

What if we won’t do that? What’s the difference?

**Comment by corpnewt** [comment]
Hey there - it seems that you've found out what the type values are for (they represent the physical port the port personality is attached to).  Yes, type 0 is for USB 2 Type-A, and type 3 is for USB 3 Type-A ports.  Worth noting that USBMap also has a `T` menu in the `P. Edit & Create USBMap.kext` menu which displays the same table as USBMapInjectorEdit (relevant code linked [here](https://github.com/corpnewt/USBMap/blob/master/USBMap.py#L833-L857)).

As for *why* we set the types - macOS uses that information to determine whether things can be hot-plugged (plugged in after boot), and setting proper port types can help with some sleep issues.

Regarding documentation - I've been trying to get the drive to write a full fledged guide on the topic - but there's tons to cover, and I haven't gotten around to it yet.  I know it's difficult to locate at times, and there are some cyclic sources out there.  If you have other questions, feel free to ask here, and if I know the answers I'll chime in though.

-CorpNewt

**Comment by istadem2077** [comment]
> I’m going to update it to `255` right now, just in case. I assume this part is not outdated.

Did it work? Or should I leave it as 3?
